src/AppInstallerRepositoryCore/Rest/RestSource.cpp

```diff
@@ -0,0 +1,262 @@
+        struct AvailablePackage : public PackageBase, public IPackage
+        {
+            using PackageBase::PackageBase;
+
+            // Inherited via IPackage
+            Utility::LocIndString GetProperty(PackageProperty property) const override
+            {
+                return PackageBase::GetProperty(property);
+            }
+
+            std::shared_ptr<IPackageVersion> GetInstalledVersion() const override
+            {
+                return {};
+            }
+
+            std::vector<PackageVersionKey> GetAvailableVersionKeys() const override
+            {
+                std::shared_ptr<const RestSource> source = GetReferenceSource();
+
+                std::vector<PackageVersionKey> result;
+                for (const auto& versionInfo : m_package.versions)
+                {
+                    result.emplace_back(
+                        source->GetIdentifier(), versionInfo.GetVersion().ToString(), versionInfo.GetChannel().ToString());
+                }
+
+                return result;
+            }
+
+            std::shared_ptr<IPackageVersion> GetLatestAvailableVersion() const override
+            {
+                return GetLatestVersionInternal();
+            }
+
+            std::shared_ptr<IPackageVersion> GetAvailableVersion(const PackageVersionKey& versionKey) const override
+            {
+                std::shared_ptr<const RestSource> source = GetReferenceSource();
+
+                // Ensure that this key targets this (or any) source
+                if (!versionKey.SourceId.empty() && versionKey.SourceId != source->GetIdentifier())
+                {
+                    return {};
+                }
+
+                std::shared_ptr<PackageVersion> packageVersion;
+                if (!versionKey.Version.empty() && !versionKey.Channel.empty())
+                {
+                    for (const auto& versionInfo : m_package.versions)
+                    {
+                        if (CaseInsensitiveEquals(versionInfo.GetVersion().ToString(), versionKey.Version)
+                            && CaseInsensitiveEquals(versionInfo.GetChannel().ToString(), versionKey.Channel))
+                        {
+                            packageVersion = std::make_shared<PackageVersion>(source, m_package.packageInfo, versionInfo);
+                        }
+                    }
+                }
+                else if (versionKey.Version.empty() && versionKey.Channel.empty())
+                {
+                    return GetLatestAvailableVersion();
+                }
+                else if (versionKey.Version.empty())
+                {
+                    for (const auto& versionInfo : m_package.versions)
+                    {
+                        if (CaseInsensitiveEquals(versionInfo.GetChannel().ToString(), versionKey.Channel))
+                        {
+                            packageVersion = std::make_shared<PackageVersion>(source, m_package.packageInfo, versionInfo);

```

Comment:
If the versions are already sorted, this will work.  If not, this just gets the first version for a channel in the list.

Also, break after